ingenialink/canopen/network.py

```diff
@@ -356,375 +359,323 @@ def load_firmware(
-                logger.error("Error detected could not specify the drive.")
-                if self._connection is not None and not servo_connected:
-                    self._teardown_connection()
-                    logger.error("CANopen connection disconnected")
-        except Exception as e:
-            logger.error("Failed to load firmware. Exception: {}".format(e))
-            error_detected_msg = "Failed to load firmware"
-            if servo is not None and not servo_connected:
-                self.disconnect_from_slave(servo)
-            elif self._connection is not None and not servo_connected:
-                self._teardown_connection()
-                logger.error("CANopen connection disconnected")
+            servo.read(PROG_STAT_1, subnode=0)
+        except ILError as e:
+            raise ILFirmwareLoadError(
+                "Firmware and bootloader versions are not compatible. Use FTP Bootloader instead."
+            ) from e
+        return servo
+
+    def __force_boot(
+        self, servo: CanopenServo, callback_status_msg: Optional[Callable[[str], None]]
+    ) -> None:
+        """Force boot, if drive is already in boot, do nothing
+
+        Args:
+            servo: target drive
+            callback_status_msg: Subscribed callback function for the status message
+
+        """
+        device_type = servo.read(CIA301_DRV_ID_DEVICE_TYPE, subnode=0)
+        device_type = device_type & 0xFFFF
+        if device_type == APPLICATION_LOADED_STATE:
+            if callback_status_msg:
+                callback_status_msg("Entering Bootmode")
+            logger.info("Entering Bootmode")
+            # Drive profile
+            # Enter in NMT pre-operational state.
+            self._connection.nmt.send_command(PROG_CTRL_STATE_FLASH)
+            # The drive will unlock the clear program command
+            password = 0x70636675
+            servo.write(FORCE_BOOT, password, subnode=0)
+
+    def __program_control_to_flash(
+        self,
+        initial_status: int,
+        servo: CanopenServo,
+        callback_status_msg: Optional[Callable[[str], None]],
+    ) -> None:
+        """Change program control status to flash.
+
+        Args:
+            initial_status: program control initial status
+            servo: target drive
+            callback_status_msg: Subscribed callback function for the status message
+
+        Raises:
+            ILFirmwareLoadError: Program control status does not change
+
+        """
+        if callback_status_msg:
+            callback_status_msg("Setting up drive")
+        target_status_list = [
+            PROG_CTRL_STATE_START,
+            PROG_CTRL_STATE_STOP,
+            PROG_CTRL_STATE_CLEAR,
+            PROG_CTRL_STATE_FLASH,
+        ]
+        status_index = target_status_list.index(initial_status) + 1
+        logger.info("Clearing program...")
+        for target_status_index in range(status_index, len(target_status_list)):
```

Comment:
You could do:
`for target_status_index, target_status in enumerate(target_status_list[status_index:], start=status_index):`